File: src/web/lib/login-experience.ts
export type LoginCompositionVariant = "BlueprintSplit" | "MonolithFocus" | "WarmMinimal";
export type LoginSurfaceTone = "WarmStone" | "WarmIvory" | "DeepNavy";
export type LoginAccentPalette = "CorporateNavyGreen" | "NavySilver" | "GreenStone";
export type LoginMotionPolicy = "Calm" | "Balanced" | "Expressive";

export interface LoginExperienceDescriptor {
  readonly version: number;
  readonly fallbackUsed: boolean;
  readonly compositionVariant: LoginCompositionVariant;
  readonly surfaceTone: LoginSurfaceTone;
  readonly accentPalette: LoginAccentPalette;
  readonly motionPolicy: LoginMotionPolicy;
  readonly eyebrow: string;
  readonly headline: string;
  readonly supportingText: string;
  readonly logoUrl: string | null;
  readonly heroUrl: string | null;
}

const compositions = ["BlueprintSplit", "MonolithFocus", "WarmMinimal"] as const;
const surfaces = ["WarmStone", "WarmIvory", "DeepNavy"] as const;
const accents = ["CorporateNavyGreen", "NavySilver", "GreenStone"] as const;
const motions = ["Calm", "Balanced", "Expressive"] as const;

export function fallbackLoginExperience(): LoginExperienceDescriptor {
  return {
    version: 0,
    fallbackUsed: true,
    compositionVariant: "BlueprintSplit",
    surfaceTone: "WarmStone",
    accentPalette: "CorporateNavyGreen",
    motionPolicy: "Balanced",
    eyebrow: "سامانه جامع مدیریت پروژه",
    headline: "ساختن، فراتر از امروز",
    supportingText: "مرکز فرمان یکپارچه برای تصمیم‌های دقیق، قابل ردیابی و مبتنی بر واقعیت پروژه.",
    logoUrl: null,
    heroUrl: null,
  };
}

export function normalizeLoginExperience(value: unknown): LoginExperienceDescriptor {
  if (!isRecord(value)) return fallbackLoginExperience();
  const version = Number(value.version);
  const compositionVariant = allowlisted(value.compositionVariant, compositions);
  const surfaceTone = allowlisted(value.surfaceTone, surfaces);
  const accentPalette = allowlisted(value.accentPalette, accents);
  const motionPolicy = allowlisted(value.motionPolicy, motions);
  const eyebrow = plainText(value.eyebrow, 80);
  const headline = plainText(value.headline, 140);
  const supportingText = plainText(value.supportingText, 320);
  if (!Number.isSafeInteger(version) || version < 0 || !compositionVariant || !surfaceTone ||
      !accentPalette || !motionPolicy || !eyebrow || !headline || !supportingText) {
    return fallbackLoginExperience();
  }

  const fallbackUsed = value.fallbackUsed === true || version === 0;
  return {
    version,
    fallbackUsed,
    compositionVariant,
    surfaceTone,
    accentPalette,
    motionPolicy,
    eyebrow,
    headline,
    supportingText,
    logoUrl: !fallbackUsed && safeAssetMarker(value.logoUrl)
      ? `/api/login-experience/assets/logo?version=${version}`
      : null,
    heroUrl: !fallbackUsed && safeAssetMarker(value.heroUrl)
      ? `/api/login-experience/assets/hero?version=${version}`
      : null,
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function allowlisted<T extends string>(value: unknown, allowed: readonly T[]): T | null {
  return typeof value === "string" && allowed.includes(value as T) ? value as T : null;
}

function plainText(value: unknown, maximumLength: number): string | null {
  if (typeof value !== "string") return null;
  const normalized = value.trim();
  return normalized.length > 0 && normalized.length <= maximumLength &&
    !/[<>\u0000-\u001f\u007f]/u.test(normalized)
    ? normalized
    : null;
}

function safeAssetMarker(value: unknown): boolean {
  return typeof value === "string" &&
    /^\/api\/v1\/public\/login-experience\/assets\/(?:logo|hero)\?/u.test(value);
}

File: src/backend/Pmcs.Modules.IdentityAccess/Domain/LoginExperience.cs
using Pmcs.BuildingBlocks.Domain;

namespace Pmcs.Modules.IdentityAccess.Domain;

public sealed class LoginExperience : AggregateRoot
{
    private LoginExperience()
    {
    }

    public Guid Id { get; private set; }

    public Guid TenantId { get; private set; }

    public int VersionNumber { get; private set; }

    public LoginExperienceStatus Status { get; private set; }

    public LoginCompositionVariant CompositionVariant { get; private set; }

    public LoginSurfaceTone SurfaceTone { get; private set; }

    public LoginAccentPalette AccentPalette { get; private set; }

    public LoginMotionPolicy MotionPolicy { get; private set; }

    public string Eyebrow { get; private set; } = string.Empty;

    public string Headline { get; private set; } = string.Empty;

    public string SupportingText { get; private set; } = string.Empty;

    public Guid? LogoDocumentId { get; private set; }

    public Guid? HeroDocumentId { get; private set; }

    public Guid CreatedBy { get; private set; }

    public DateTimeOffset CreatedAt { get; private set; }

    public Guid? PublishedBy { get; private set; }

    public DateTimeOffset? PublishedAt { get; private set; }

    public static LoginExperience CreateDraft(
        Guid id,
        Guid tenantId,
        int versionNumber,
        LoginCompositionVariant compositionVariant,
        LoginSurfaceTone surfaceTone,
        LoginAccentPalette accentPalette,
        LoginMotionPolicy motionPolicy,
        string eyebrow,
        string headline,
        string supportingText,
        Guid? logoDocumentId,
        Guid? heroDocumentId,
        Guid createdBy,
        DateTimeOffset createdAt)
    {
        if (id == Guid.Empty || tenantId == Guid.Empty || createdBy == Guid.Empty)
        {
            throw new DomainRuleException(
                "login-experience.identity.required",
                "Login experience, tenant and creator ids are required.");
        }

        if (versionNumber <= 0)
        {
            throw new DomainRuleException(
                "login-experience.version.invalid",
                "Login experience version must be positive.");
        }

        ValidateEnum(compositionVariant, "composition");
        ValidateEnum(surfaceTone, "surface tone");
        ValidateEnum(accentPalette, "accent palette");
        ValidateEnum(motionPolicy, "motion policy");
        ValidateAssetId(logoDocumentId);
        ValidateAssetId(heroDocumentId);

        return new LoginExperience
        {
            Id = id,
            TenantId = tenantId,
            VersionNumber = versionNumber,
            Status = LoginExperienceStatus.Draft,
            CompositionVariant = compositionVariant,
            SurfaceTone = surfaceTone,
            AccentPalette = accentPalette,
            MotionPolicy = motionPolicy,
            Eyebrow = RequiredPlainText(eyebrow, 80, "login-experience.eyebrow.invalid"),
            Headline = RequiredPlainText(headline, 140, "login-experience.headline.invalid"),
            SupportingText = RequiredPlainText(
                supportingText,
                320,
                "login-experience.supporting-text.invalid"),
            LogoDocumentId = logoDocumentId,
            HeroDocumentId = heroDocumentId,
            CreatedBy = createdBy,
            CreatedAt = createdAt
        };
    }

    public bool MatchesDraft(
        LoginCompositionVariant compositionVariant,
        LoginSurfaceTone surfaceTone,
        LoginAccentPalette accentPalette,
        LoginMotionPolicy motionPolicy,
        string eyebrow,
        string headline,
        string supportingText,
        Guid? logoDocumentId,
        Guid? heroDocumentId) =>
        CompositionVariant == compositionVariant &&
        SurfaceTone == surfaceTone &&
        AccentPalette == accentPalette &&
        MotionPolicy == motionPolicy &&
        string.Equals(Eyebrow, eyebrow.Trim(), StringComparison.Ordinal) &&
        string.Equals(Headline, headline.Trim(), StringComparison.Ordinal) &&
        string.Equals(SupportingText, supportingText.Trim(), StringComparison.Ordinal) &&
        LogoDocumentId == logoDocumentId &&
        HeroDocumentId == heroDocumentId;

    public void Publish(long baseRevision, Guid publishedBy, DateTimeOffset publishedAt)
    {
        RequireRevision(baseRevision);
        if (Status is not LoginExperienceStatus.Draft and not LoginExperienceStatus.Superseded)
        {
            throw new DomainRuleException(
                "login-experience.publish.invalid-state",
                "Only a draft or superseded login experience can be published.");
        }

        if (publishedBy == Guid.Empty)
        {
            throw new DomainRuleException(
                "login-experience.publisher.required",
                "A publisher is required.");
        }

        Status = LoginExperienceStatus.Published;
        PublishedBy = publishedBy;
        PublishedAt = publishedAt;
        AdvanceRevision();
    }

    public void MarkSuperseded(Guid actorUserId, DateTimeOffset changedAt)
    {
        if (Status != LoginExperienceStatus.Published)
        {
            return;
        }

        if (actorUserId == Guid.Empty)
        {
            throw new DomainRuleException(
                "login-experience.actor.required",
                "A superseding actor is required.");
        }

        Status = LoginExperienceStatus.Superseded;
        PublishedBy = actorUserId;
        PublishedAt = changedAt;
        AdvanceRevision();
    }

    private void RequireRevision(long baseRevision)
    {
        if (Revision != baseRevision)
        {
            throw new DomainRuleException(
                "login-experience.revision.conflict",
                "The login experience changed after it was loaded.");
        }
    }

    private static string RequiredPlainText(string value, int maximumLength, string code)
    {
        var normalized = value?.Trim() ?? string.Empty;
        if (normalized.Length == 0 || normalized.Length > maximumLength ||
            normalized.Any(char.IsControl) || normalized.Contains('<') || normalized.Contains('>'))
        {
            throw new DomainRuleException(code, "Login presentation text is outside the allowlisted policy.");
        }

        return normalized;
    }

    private static void ValidateAssetId(Guid? value)
    {
        if (value == Guid.Empty)
        {
            throw new DomainRuleException(
                "login-experience.asset.invalid",
                "Login experience asset id cannot be empty.");
        }
    }

    private static void ValidateEnum<TEnum>(TEnum value, string label)
        where TEnum : struct, Enum
    {
        if (!Enum.IsDefined(value))
        {
            throw new DomainRuleException(
                "login-experience.token.invalid",
                $"Login experience {label} is not allowlisted.");
        }
    }
}

public enum LoginExperienceStatus
{
    Draft = 1,
    Published = 2,
    Superseded = 3
}

public enum LoginCompositionVariant
{
    BlueprintSplit = 1,
    MonolithFocus = 2,
    WarmMinimal = 3
}

public enum LoginSurfaceTone
{
    WarmStone = 1,
    WarmIvory = 2,
    DeepNavy = 3
}

public enum LoginAccentPalette
{
    CorporateNavyGreen = 1,
    NavySilver = 2,
    GreenStone = 3
}

public enum LoginMotionPolicy
{
    Calm = 1,
    Balanced = 2,
    Expressive = 3
}
